### ISSUE TITLE
Error when validating string against object schema

### ISSUE THREAD

**Issue by bausmeier** [jonasschmidt/ex_json_schema#5]
I get the following error when validating a string against a schema with type "object":

```
** (BadMapError) expected a map, got: "boom"
    (stdlib) :maps.is_key("name", "boom")
    lib/ex_json_schema/validator.ex:116: anonymous fn/2 in ExJsonSchema.Validator.validate_aspect/4
    (elixir) lib/enum.ex:873: anonymous fn/3 in Enum.flat_map/2
    (elixir) lib/enum.ex:1473: Enum."-reduce/3-lists^foldl/2-0-"/3
    (elixir) lib/enum.ex:872: Enum.flat_map/2
    (elixir) lib/enum.ex:873: anonymous fn/3 in Enum.flat_map/2
    (stdlib) lists.erl:1262: :lists.foldl/3
    (elixir) lib/enum.ex:872: Enum.flat_map/2
```

This happens when it tries to validate the required keys of the object when it is in fact a string. I expect that the type should be checked first and if it is invalid then no further validation should be done. Does that sound reasonable?

I'm happy to have a go at fixing this in a PR, I just want to understand what the correct behaviour is first.

**Comment by jonasschmidt** [comment]
Yes, this does indeed sound like something that should be handled gracefully. I will look into it to find out what the expected behaviour is supposed to be and why it's not covered by one of the test cases in the official JSON Schema test suite.

**Comment by jonasschmidt** [comment]
So I looked into it a bit and there's no behaviour defined in the standard on how to deal with that. JSON Schema doesn't even tell you how to indicate that a validation failed. So an exception could also indicate a negative result of a validation.

That being said, from a library design perspective it's a bug of course. I will have a fix ready today or a bit later this week.

**Comment by bausmeier** [comment]
Great, thanks. The fix isn't urgent for me so no rush.

**Comment by bausmeier** [comment]
Thanks for the fix - it works well now.

**Comment by jonasschmidt** [comment]
Great. It's released as `v0.4.0` as well.
